lgc/builder/ImageBuilder.cpp

```diff
@@ -1311,14 +1312,13 @@ Value *ImageBuilder::CreateImageQuerySamples(unsigned dim, unsigned flags, Value
   // Sample number = 1 << LAST_LEVEL
   Value *sampleNumber = CreateShl(getInt32(1), lastLevel);
 
-  // Extract TYPE(SQ_IMG_RSRC_WORD3, [31:28])
-  Value *imageType = CreateIntrinsic(Intrinsic::amdgcn_ubfe, getInt32Ty(), {descWord3, getInt32(28), getInt32(4)});
-
-  // Check if resource type is 2D MSAA or 2D MSAA array, 14 = SQ_RSRC_IMG_2D_MSAA, 15 = SQ_RSRC_IMG_2D_MSAA_ARRAY
-  Value *isMsaa = CreateOr(CreateICmpEQ(imageType, getInt32(14)), CreateICmpEQ(imageType, getInt32(15)));
-
-  // Return sample number if resource type is 2D MSAA or 2D MSAA array. Otherwise, return 1.
-  return CreateSelect(isMsaa, sampleNumber, getInt32(1), instName);
+  // Extract DATA_FORMAT (SQ_IMG_RSRC_WORD1, [25:20]
+  Value *descWord1 = CreateExtractElement(imageDesc, 1);
+  Value *dataFormat = CreateIntrinsic(Intrinsic::amdgcn_ubfe, getInt32Ty(), {descWord1, getInt32(20), getInt32(6)});
```

Comment:
Sorry for not follow your suggeston. Because the internal compiler don't have this issue, I thought we might follow the way of the internal compiler. I will make another change as you suggested.

Reply:
Fixed. Please reivew.

Reply:
Why use ubfe with constant arguments instead of shift and and? I think instcombine will break this anyway

Reply:
Hi @arsenm, the change has been updated. The current change will clamp the sample number to 0 if dword3 is equal to 0 and allowNullDescriptor is enabled.